controllers/search.js

```diff
@@ -0,0 +1,175 @@
+const articleAttrib = ['title', 'description', 'slug', 'tagList'];
+
+const handler = {
+  searchByAuthor: async (req, res) => {
+    const { title, username } = req.body;
+
+    if (!title && !username) {
+      return res.status(400).send({
+        status: res.statusCode,
+        error: 'Please search using the article title or authors\' username'
+      });
+    }
+
+    try {
+      /**
+      * Any user is able to search for an article if
+      * they know it's tittle
+      */
+
+      if (title) {
+        const articles = await article.findAll(
+          {
+            limit: 10,
+            attributes: articleAttrib,
+            where: {
+              title: { [Sequelize.Op.iLike]: `%${title}%` }
+            },
+            include: {
+              model: User,
+              attributes: userAttrib,
+            }
+          }
+        );
+        return res.status(201).send({
+          status: res.statusCode,
+          articles
+        });
+      }
+
+      /**
+      * Any user is able to search for an article if
+      * they know the authors name
+      */
+
+      if (username) {
+        const users = await User.findAll({
+          where: {
+            username: { [Sequelize.Op.iLike]: `%${username}%` }
+          },
+          attributes: userAttrib,
+          include: [{
+            model: article,
+            attributes: articleAttrib,
+          }],
+          limit: 10,
+        });
+        return res.status(201).send({
+          status: res.statusCode,
+          users
+        });
+      }
+    } catch (error) {
+      return res.status(500).send({
+        status: res.statusCode,
+        error: 'Server failed to handle your request'
+
+      });
+    }
+  },
+
+  searchByBody: async (req, res) => {
+    const { body } = req.body;
+    try {
+      const articles = await article.findAll({
+        where: {
+          body: {
+            [Sequelize.Op.iLike]: `%${body}%`
+          }
+        },
+        attributes: articleAttrib,
+      });
+      return res.status(200).send({
+        status: res.statusCode,
+        articles,
+      });
+    } catch (error) {
+      return res.status(500).send({
+        status: res.statusCode,
+        error: 'Failed to handle your request',
+      });
+    }
+  },
+
+  searchByTag: async (req, res) => {
+    const { tag } = req.body;
+    try {
+      const articles = await article.findAll({
+        where: {
+          tagList: {
+            [Sequelize.Op.contains]: tag.toLowerCase(),
+          }
+        },
+        attributes: articleAttrib,
+      });
+      return res.status(200).send({
+        status: res.statusCode,
+        articles,
+      });
+    } catch (error) {
+      return res.status(500).send({
+        status: res.statusCode,
+        error: 'Failed to handle your request',
+      });
+    }
+  },
+
+  searchAll: async (req, res) => {
+    const { all: query } = req.body;
+    try {
+      const users = await User.findAll({
```

Comment:
This too will return all users with that username even if they don't have an article! as commented on line 54.

Reply:
@avpaul  The logic I wanted to implement here as I explained to you [In person] it's like when someone makes a general search the application has to return any matching criteria (article, the user or tag)[example](https://medium.com/search?q=bobo)